### DISCUSSION TITLE
dynamic colourmap style of tiff

### DISCUSSION THREAD

**Discussion by hardreddata** [TerriaJS/terriajs#6417] [General]
I have a big Digital Elevation Model, and if I style the whole thing based on the range of the data it lacks the detail in certain places.

I was thinking it might be quite nice to style it dynamically based on viewing extent.

I wondered if anyone had any successes with this or similar?

There is https://docs.geoserver.org/stable/en/user/extensions/sldservice/index.html but I am not sure Terria is equipped to call this service and feed the style back into the WMS request.

I also found https://docs.geoserver.org/stable/en/user/community/colormap/index.html

I thought perhaps we might have something along these lines for Digital Earth Australia?

Advice invited.

Thanks.

**Comment by nf-s** [comment]
Hi @RussellGrew 

I have had some success using the SLD [variable substitution](https://docs.geoserver.org/stable/en/user/styling/sld/extensions/substitution.html) to dynamically set min/max values for raster color scales (using parameters in `GetMap` requests). Unfortunately I lost access to that GeoServer (when I moved jobs), so I can't dig out the styles.

I don't have any experience with the sldservice or colormap extensions. I can't think of an easy way that Terria could interact with the sldservice.

Another Geoserver extension is ncMWS - https://docs.geoserver.org/stable/en/user/community/ncwms/index.html - which I haven't used before but looks promising.

Terria supports ncWMS, so it will automatically add the color scale range to the workbench - http://ci.terria.io/main/#share=s-AkYXjX29wAbvJOLUl6gIKFLDVf

Are you looking for the style to change automatically as you interact with the map?
Or just to manually change the style's min/max value?

@staffordsmith83 may have something to add here - he has done a bit with GeoServer dynamic styling.

Digital Earth Australia is serving WMS through https://github.com/opendatacube/datacube-ows
I am not that familiar with the codebase so I can't comment on how they do styling

Cheers,
Nick

**Comment by hardreddata** [comment]
Hi @nf-s 

Thanks for the advice.

I don't think I can _get_ raster min and max inside my map extent using SLD substitution though I see you could set them for the style.

I was not aware of the ncWMS community module. This looks promising. Maybe there are complications with GeoServer community modules when compared to extensions, as they aren't built to align with release versions of GeoServer.

There are some details here https://stackoverflow.com/questions/59654177/geoserver-ncwms-extension-gettimeseries-doesnt-work

That aside I have it working, for manual minimum and maximum and a custom colour bar.

I had assumed the dynamic case from map extents was easier but was not aware of the nice ncWMS support in TerriaJS.

I will let you know if I have any wins.

**Comment by frizwi** [comment]
Hi,
Just wondering if anyone has a workaround for images that are served up by ncWMS2/THREDDS? There is a min/max colour range that has to be manually edited, but Godiva provides an autoscale button which scales the image based on the current view - is this possible to do in Terria? Ideally dynamically as you pan/zoom but an explicit button will do just fine.
Thanks!
-Farhan
